builtin/providers/aws/resource_aws_lambda_function.go

```diff
@@ -287,7 +279,96 @@ func resourceAwsLambdaFunctionDelete(d *schema.ResourceData, meta interface{}) e
 // resourceAwsLambdaFunctionUpdate maps to:
 // UpdateFunctionCode in the API / SDK
 func resourceAwsLambdaFunctionUpdate(d *schema.ResourceData, meta interface{}) error {
-	return nil
+	conn := meta.(*AWSClient).lambdaconn
+
+	d.Partial(true)
+
+	codeReq := &lambda.UpdateFunctionCodeInput{
+		FunctionName: aws.String(d.Id()),
+	}
+
+	codeUpdate := false
+	if v, ok := d.GetOk("filename"); ok && d.HasChange("source_code_hash") {
+		file, err := loadFileContent(v.(string))
+		if err != nil {
+			return fmt.Errorf("Unable to load %q: %s", v.(string), err)
+		}
+		codeReq.ZipFile = file
+		codeUpdate = true
+	}
+	if d.HasChange("s3_bucket") || d.HasChange("s3_key") || d.HasChange("s3_object_version") {
+		d.SetPartial("s3_bucket")
+		d.SetPartial("s3_key")
+		d.SetPartial("s3_object_version")
```

Comment:
Right, I think it does - basically it means I'll need to move all of the `d.SetPartial(` calls below `conn.UpdateFunctionCode` so these are only called when `conn.UpdateFunctionCode` is actually successful.

The same will apply to `UpdateFunctionConfiguration`.